rust/lance/src/io/commit/conflict_resolver.rs

```diff
@@ -827,13 +1002,152 @@ impl<'a> TransactionRebase<'a> {
+                        || (added.is_empty() && updated.is_empty() && !removed.is_empty())
+                    {
+                        return Ok(());
+                    }
+
+                    // 2. MemWALs of different regions can be changed at the same time
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_added,
+                        added,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_added,
+                        updated,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_updated,
+                        added,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    self.check_update_mem_wal_state_not_modify_same_mem_wal(
+                        committed_updated,
+                        updated,
+                        other_transaction,
+                        other_version,
+                    )?;
+                    Ok(())
+                }
+                Operation::Update {
+                    mem_wal_to_flush, ..
+                } => {
+                    if mem_wal_to_flush.is_some() {
+                        Ok(())
+                    } else {
+                        Err(self.incompatible_conflict_err(
+                            other_transaction,
+                            other_version,
+                            location!(),
+                        ))
+                    }
+                }
+                Operation::UpdateConfig { .. }
+                | Operation::Rewrite { .. }
+                | Operation::CreateIndex { .. }
+                | Operation::ReserveFragments { .. } => Ok(()),
+                Operation::Append { .. }
+                | Operation::Overwrite { .. }
+                | Operation::Delete { .. }
+                | Operation::DataReplacement { .. }
+                | Operation::Merge { .. }
+                | Operation::Restore { .. }
+                | Operation::Project { .. } => Err(self.incompatible_conflict_err(
+                    other_transaction,
+                    other_version,
+                    location!(),
+                )),
+            }
+        } else {
+            Err(wrong_operation_err(&self.transaction.operation))
+        }
+    }
+
+    fn check_update_mem_wal_state_not_modify_same_mem_wal(
+        &self,
+        committed: &[MemWal],
+        to_commit: &[MemWal],
+        other_transaction: &Transaction,
+        other_version: u64,
+    ) -> Result<()> {
+        if !committed.is_empty() {
+            if committed.is_empty() || to_commit.is_empty() {
```

Comment:
The `committed.is_empty()` part here seems redundant.  It must be false or we wouldn't be here.